circuits/kimchi/src/polynomials/poseidon.rs

```diff
@@ -78,96 +87,28 @@ pub fn constraint<F: FftField + SquareRootField>() -> E<F> {
         })
         .collect();
 
-    for e in ROUND_EQUATIONS.iter() {
-        let &RoundEquation {
-            source,
-            target: (target_row, target_round),
-        } = e;
-        let sboxed: Vec<_> = round_to_cols(source)
-            .map(|i| {
-                cache.cache(
-                    E::cell(Column::Witness(i), Curr).pow(PlonkSpongeConstants15W::SPONGE_BOX),
-                )
-            })
-            .collect();
-
-        res.extend(round_to_cols(target_round).enumerate().map(|(j, col)| {
-            let rc = E::cell(Column::Coefficient(idx), Curr);
-
-            idx += 1;
-
-            E::cell(Column::Witness(col), target_row)
-                - sboxed
-                    .iter()
-                    .zip(mds[j].iter())
-                    .fold(rc, |acc, (x, c)| acc + E::Constant(c.clone()) * x.clone())
-        }));
-    }
-    E::cell(Column::Index(GateType::Poseidon), Curr) * E::combine_constraints(0, res)
-}
-
-impl<F: FftField + SquareRootField> ConstraintSystem<F> {
-    pub fn psdn_scalars(
-        evals: &[ProofEvaluations<F>],
-        params: &ArithmeticSpongeParams<F>,
-        alpha: &[F],
-    ) -> Vec<F> {
-        let w_zeta = evals[0].w;
-        let sboxed: [[F; SPONGE_WIDTH]; ROUNDS_PER_ROW] = array_init(|round| {
-            array_init(|i| {
-                let col = round_to_cols(round);
-                sbox::<F, PlonkSpongeConstants15W>(w_zeta[col][i])
-            })
-        });
-        let alp: [[F; SPONGE_WIDTH]; ROUNDS_PER_ROW] =
-            array_init(|round| array_init(|i| alpha[round * SPONGE_WIDTH + i]));
-
-        let lhs = ROUND_EQUATIONS.iter().fold(F::zero(), |acc, eq| {
-            let (target_row, target_round) = &eq.target;
-            let cols = match target_row {
-                Curr => &evals[0].w,
-                Next => &evals[1].w,
-            };
-            cols[round_to_cols(*target_round)]
-                .iter()
-                .zip(alp[eq.source].iter())
-                .map(|(p, a)| -*a * p)
-                .fold(acc, |x, y| x + y)
-        });
-
-        let mut rhs = F::zero();
-        for eq in ROUND_EQUATIONS.iter() {
-            let ss = sboxed[eq.source];
-            let aa = alp[eq.source];
-            for (i, p) in ss.iter().enumerate() {
-                // Each of these contributes to the right hand side of SPONGE_WIDTH cell equations
-                let coeff =
-                    (0..SPONGE_WIDTH).fold(F::zero(), |acc, j| acc + aa[j] * params.mds[j][i]);
-                rhs += coeff * p;
-            }
+    for (round, eq) in ROUND_EQUATIONS.iter().enumerate() {
```

Comment:
The existing implementation was written carefully to ensure efficiency. If you want to include this rewrite can you make sure there was no degradation? If it's difficult to tell please revert otherwise

Reply:
the refactor is doing exactly the same thing (just in a cleaner way), unless I made a mistake somewhere,

Reply:
let me double check by printing though